File: cses/02_sorting_searching/2428_distinct-values-subarrays-ii/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    k = int(data[1])
    # Giữ nguyên dạng token bytes: vừa hashable vừa so sánh được, tránh chi phí parse int.
    a = data[2:2 + n]

    # Nén tọa độ về [0, n) để dùng mảng đếm tần suất bằng list (nhanh hơn dict).
    comp = {}
    idx = 0
    arr = [0] * n
    for i in range(n):
        v = a[i]
        j = comp.get(v)
        if j is None:
            j = idx
            comp[v] = j
            idx += 1
        arr[i] = j

    # Two-pointer / sliding window: duy trì cửa sổ [left, right] có <= k giá trị phân biệt.
    cnt = [0] * idx
    distinct = 0
    left = 0
    total = 0
    for right in range(n):
        # Thêm phần tử bên phải; nếu tần suất cũ bằng 0 thì có thêm một giá trị mới.
        v = arr[right]
        if cnt[v] == 0:
            distinct += 1
        cnt[v] += 1
        # Thu nhỏ cửa sổ từ bên trái cho tới khi số giá trị phân biệt <= k.
        while distinct > k:
            lv = arr[left]
            cnt[lv] -= 1
            if cnt[lv] == 0:
                distinct -= 1
            left += 1
        # Mọi subarray kết thúc tại right, bắt đầu trong [left, right] đều hợp lệ.
        total += right - left + 1

    sys.stdout.write(str(total) + "\n")


main()

File: cses/02_sorting_searching/2428_distinct-values-subarrays-ii/solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, maximumDistinct;
    cin >> n >> maximumDistinct;

    // Nén tọa độ: ánh xạ mỗi giá trị gốc về một chỉ số trong [0, distinctValues)
    // để về sau dùng mảng đếm tần suất (truy cập theo chỉ số nhanh hơn hash map).
    vector<int> values(n);
    unordered_map<int, int> compressedIndex;
    compressedIndex.reserve(2 * n);
    compressedIndex.max_load_factor(0.7F);
    int distinctValues = 0;
    for (int& value : values) {
        int original;
        cin >> original;
        const auto [iterator, inserted] =
            compressedIndex.emplace(original, distinctValues);
        if (inserted) {
            ++distinctValues;
        }
        value = iterator->second;
    }

    // Two-pointer / sliding window: duy trì cửa sổ [left, right] có <= k giá trị phân biệt.
    vector<int> frequency(distinctValues, 0);
    int windowDistinct = 0;
    int left = 0;
    long long answer = 0;

    for (int right = 0; right < n; ++right) {
        // Thêm phần tử bên phải; nếu trước đó tần suất bằng 0 thì có thêm một giá trị mới.
        if (frequency[values[right]]++ == 0) {
            ++windowDistinct;
        }
        // Thu nhỏ cửa sổ từ bên trái cho tới khi số giá trị phân biệt <= k.
        while (windowDistinct > maximumDistinct) {
            if (--frequency[values[left]] == 0) {
                --windowDistinct;
            }
            ++left;
        }
        // Mọi subarray kết thúc tại right, bắt đầu trong [left, right] đều hợp lệ.
        answer += right - left + 1;
    }

    cout << answer << '\n';
    return 0;
}
